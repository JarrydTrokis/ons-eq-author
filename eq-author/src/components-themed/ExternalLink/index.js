import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const IconContainer = styled.span`
  visibility: hidden;
  white-space: nowrap;
`;

const ExternalIcon = styled.svg`
  vertical-align: middle;
  width: 1rem;
  height: 1rem;
  padding-bottom: 0.1rem;
  margin: 0 0 0 0.25rem;
  visibility: visible;
`;

const VisualHidden = styled.span`
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1px;
`;

const ExternalLink = ({ url, linkText }) => {
  return (
    <a
      href={url}
      target="_blank"
      rel="noopener noreferrer"
      data-test="external-link-href"
    >
      <span data-test="external-link-text">{linkText}</span>
      <IconContainer>
        &nbsp;
        <ExternalIcon
          id="external-link"
          viewBox="0 0 12 12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13.5,9H13a.5.5,0,0,0-.5.5v3h-9v-9h3A.5.5,0,0,0,7,3V2.5A.5.5,0,0,0,6.5,2h-4a.5.5,0,0,0-.5.5v11a.5.5,0,0,0,.5.5h11a.5.5,0,0,0,.5-.5v-4A.5.5,0,0,0,13.5,9Z"
            transform="translate(-2 -1.99)"
          />
          <path
            d="M8.83,7.88a.51.51,0,0,0,.71,0l2.31-2.32,1.28,1.28A.51.51,0,0,0,14,6.49v-4a.52.52,0,0,0-.5-.5h-4A.51.51,0,0,0,9,2.52a.58.58,0,0,0,.14.33l1.28,1.28L8.12,6.46a.51.51,0,0,0,0,.71Z"
            transform="translate(-2 -1.99)"
          />
        </ExternalIcon>
      </IconContainer>
      <VisualHidden data-test="external-link-screen-reader-text">
        {" "}
        (opens in a new tab)
      </VisualHidden>
    </a>
  );
};

ExternalLink.propTypes = {
  url: PropTypes.string.isRequired,
  linkText: PropTypes.string.isRequired,
};

export default ExternalLink;
