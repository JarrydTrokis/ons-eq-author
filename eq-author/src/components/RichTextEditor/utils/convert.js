import { convertToHTML, convertFromHTML } from "draft-convert";
import { EditorState } from "draft-js";

export const toHTML = entityMap => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return editorState => convert(editorState.getCurrentContent());
};

export const fromHTML = nodeToFn => {
  const htmlToEntity = (nodeName, ...otherArgs) => {
    const entity = Object.entries(nodeToFn)
      .filter(([name]) => name === nodeName)
      .map(([, fn]) => fn(nodeName, ...otherArgs))
      .find(result => result);

    return entity || null;
  };

  const convert = convertFromHTML({ htmlToEntity });

  return (html, decorator) =>
    EditorState.createWithContent(convert(html), decorator);
};
