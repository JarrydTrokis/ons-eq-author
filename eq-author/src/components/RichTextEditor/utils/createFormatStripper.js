import { mapKeys } from "lodash";
import { filterEditorState } from "draftjs-filters";
import { filterConfig as pipedFilterConfig } from "components/RichTextEditor/entities/PipedValue";
import { filterConfig as linkFilterConfig } from "components/RichTextEditor/LinkPlugin";

const mapper = {
  bold: { format: "BOLD", type: "styles" },
  emphasis: { format: "ITALIC", type: "styles" },
  list: { format: "unordered-list-item", type: "blocks" },
  heading: { format: "header-two", type: "blocks" },
};

// Custom filter function to replace multiple spaces with a single space
function replaceMultipleSpaces(contentState) {
  const text = contentState.getText();
  const transformedText = text.replace(/\s+/g, " "); // Replace multiple spaces with a single space
  return contentState.merge({
    text: transformedText,
  });
}

export default function createFormatStripper(controls) {
  const filterConfiguration = {
    blocks: [],
    styles: [],
    entities: [pipedFilterConfig, linkFilterConfig],
    maxNesting: 0,
    whitespacedCharacters: [replaceMultipleSpaces],
  };

  mapKeys(controls, (value, key) => {
    if (mapper[key] && controls[key]) {
      filterConfiguration[mapper[key].type].push(mapper[key].format);
    }
  });

  return (editorState) => filterEditorState(filterConfiguration, editorState);
}
