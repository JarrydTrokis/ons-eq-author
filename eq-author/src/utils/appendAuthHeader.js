export default headers => {
  const accessToken = localStorage.getItem("accessToken");
  const returnedHeaders = {
    ...headers
  };

  if (accessToken) {
    returnedHeaders.authorization = `Bearer ${accessToken}`;
  }

  return {
    headers: {
      ...returnedHeaders
    }
  };
};
