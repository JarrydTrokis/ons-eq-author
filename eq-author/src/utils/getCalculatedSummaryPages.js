const insertSection = (sections, section, folder, page, answers) => {
  sections.push({
    ...section,
    folders: [],
  });
  sections = insertFolder(sections, folder, page, answers);
  return sections;
};

const insertFolder = (sections, folder, page, answers) => {
  sections[sections.length - 1].folders.push({
    ...folder,
    pages: [],
  });
  sections = insertPage(sections, page, answers);
  return sections;
};

const insertPage = (sections, page, answers) => {
  sections[sections.length - 1].folders[
    sections[sections.length - 1].folders.length - 1
  ].pages.push({
    ...page,
    answers,
  });
  return sections;
};

const getCalculatedSummaryPages = (questionnaire, sectionId) => {
  const sections = [];
  if (questionnaire?.sections) {
    for (const section of questionnaire?.sections) {
      if (section.id === sectionId) {
        return sections;
      }
      for (const folder of section?.folders) {
        for (const page of folder?.pages) {
          const answers = page?.answers;
          if (page?.pageType === "CalculatedSummaryPage") {
            if (sections[sections.length - 1]?.id === section?.id) {
              insertPage(sections, page, answers);
            } else {
              insertSection(sections, section, folder, page, answers);
            }
          }
        }
      }
    }
  }
  return sections;
};

export default getCalculatedSummaryPages;
