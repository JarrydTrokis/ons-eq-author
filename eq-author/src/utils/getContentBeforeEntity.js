import { remove } from "lodash";

import isListCollectorPageType from "utils/isListCollectorPageType";

import { MUTUALLY_EXCLUSIVE } from "constants/answer-types";

const identity = (x) => x;

const getContentBeforeEntity = (
  questionnaire,
  id,
  preprocessAnswers,
  includeTarget,
  expressionGroup
) => {
  const sections = [];

  for (const section of questionnaire.sections) {
    if (section.id === id) {
      return sections;
    }

    sections.push({
      ...section,
      folders: [],
    });

    for (const folder of section.folders) {
      if (folder.id === id) {
        return sections;
      }

      sections[sections.length - 1].folders.push({
        ...folder,
        pages: [],
      });

      for (const page of folder.pages) {
        if (page.id === id && !includeTarget) {
          return sections;
        }

        let answers =
          !isListCollectorPageType(page.pageType) &&
          (page?.answers?.flatMap(preprocessAnswers) || []);

        if (expressionGroup?.operator === "And") {
          expressionGroup.expressions.forEach((expression) => {
            if (expression?.left?.page?.id) {
              answers = answers.filter((answer) => {
                if (expression.left.page.id === answer.page.id) {
                  if (expression.left.type === MUTUALLY_EXCLUSIVE) {
                    return false;
                  } else {
                    return answer.type !== MUTUALLY_EXCLUSIVE;
                  }
                }
                return true;
              });
            }
          });
        }

        if (answers.length) {
          sections[sections.length - 1].folders[
            sections[sections.length - 1].folders.length - 1
          ].pages.push({
            ...page,
            answers,
          });
        }

        if (
          (page.id === id && includeTarget) ||
          page?.confirmation?.id === id
        ) {
          return sections;
        }
      }
    }
  }

  return sections;
};

export const stripEmpty = (sections) => {
  sections.forEach(({ folders }) =>
    remove(folders, (folder) => !folder.pages.length)
  );
  remove(sections, (section) => !section.folders.length);
  return sections;
};

export default ({
  questionnaire,
  id,
  preprocessAnswers = identity,
  includeTargetPage = false,
  expressionGroup,
} = {}) => {
  if (!questionnaire || !id || questionnaire?.introduction?.id === id) {
    return [];
  }

  return stripEmpty(
    getContentBeforeEntity(
      questionnaire,
      id,
      preprocessAnswers,
      includeTargetPage,
      expressionGroup
    ).filter(({ folders }) => folders.length)
  );
};
