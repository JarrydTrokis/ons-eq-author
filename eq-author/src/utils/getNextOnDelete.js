import { findIndex } from "lodash";

export default (items, id) => {
  const index = findIndex(items, { id });
  let nextIndex;

  if (index === 0) {
    nextIndex = 0;
  } else {
    nextIndex = index - 1;
  }

  return nextIndex;
};
