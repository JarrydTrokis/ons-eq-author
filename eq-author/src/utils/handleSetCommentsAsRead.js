import { useEffect } from "react";
import UPDATE_COMMENTS_AS_READ from "../graphql/updateCommentsAsRead.graphql";
import { useMutation } from "@apollo/react-hooks";

// Function begins with capital letter - allows useEffect to be used here
const HandleSetCommentsAsRead = (pageId, userId) => {
  const [updateCommentsAsRead] = useMutation(UPDATE_COMMENTS_AS_READ);

  // https://stackoverflow.com/questions/66404382/how-to-detect-route-changes-using-react-router-in-react
  useEffect(() => {
    return function cleanup() {
      updateCommentsAsRead({
        variables: {
          input: {
            pageId,
            userId,
          },
        },
      });
    };
  }, [updateCommentsAsRead, pageId, userId]);
};

export default HandleSetCommentsAsRead;
