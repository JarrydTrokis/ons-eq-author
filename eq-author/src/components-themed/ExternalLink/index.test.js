import React from "react";
import { render } from "tests/utils/rtl";

import ExternalLink from ".";

describe("Inline external link render", () => {
  it("should render as expected", () => {
    const externalLink = render(
      <ExternalLink url="https://www.google.com/" linkText="link to google" />
    );
    expect(externalLink).toMatchSnapshot();
  });
});
