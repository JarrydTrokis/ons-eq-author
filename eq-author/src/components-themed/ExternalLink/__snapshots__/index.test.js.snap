// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline external link render should render as expected 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c3 {
  list-style: none;
  margin: 0;
  padding: 0;
}

.c1 {
  position: absolute;
  bottom: 0;
  text-align: center;
  min-width: 10em;
  margin-bottom: 0;
  z-index: 99999999;
}

.c2 {
  display: inline-block;
  text-align: initial;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  visibility: hidden;
  white-space: nowrap;
}

.c5 {
  vertical-align: middle;
  width: 1rem;
  height: 1rem;
  padding-bottom: 0.1rem;
  margin: 0 0 0 0.25rem;
  visibility: visible;
}

.c6 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1px;
}

<body>
    <div
      id="toast"
    >
      <div
        class="c0"
      >
        <div
          class="c1"
        >
          <div
            class="c2"
          >
            <ol
              aria-live="assertive"
              aria-relevant="additions removals"
              class="c3"
              role="alert"
            />
          </div>
        </div>
      </div>
    </div>
    <div>
      <a
        href="https://www.google.com/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span>
          link to google
        </span>
        <span
          class="c4"
        >
           
          <svg
            class="c5"
            id="external-link"
            viewBox="0 0 12 12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.5,9H13a.5.5,0,0,0-.5.5v3h-9v-9h3A.5.5,0,0,0,7,3V2.5A.5.5,0,0,0,6.5,2h-4a.5.5,0,0,0-.5.5v11a.5.5,0,0,0,.5.5h11a.5.5,0,0,0,.5-.5v-4A.5.5,0,0,0,13.5,9Z"
              transform="translate(-2 -1.99)"
            />
            <path
              d="M8.83,7.88a.51.51,0,0,0,.71,0l2.31-2.32,1.28,1.28A.51.51,0,0,0,14,6.49v-4a.52.52,0,0,0-.5-.5h-4A.51.51,0,0,0,9,2.52a.58.58,0,0,0,.14.33l1.28,1.28L8.12,6.46a.51.51,0,0,0,0,.71Z"
              transform="translate(-2 -1.99)"
            />
          </svg>
        </span>
        <span
          class="c6"
        >
           (opens in a new tab)
        </span>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      href="https://www.google.com/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <span>
        link to google
      </span>
      <span
        class="ExternalLink__IconContainer-jbkyyd-0 iyisdy"
      >
         
        <svg
          class="ExternalLink__ExternalIcon-jbkyyd-1 ktvDYI"
          id="external-link"
          viewBox="0 0 12 12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13.5,9H13a.5.5,0,0,0-.5.5v3h-9v-9h3A.5.5,0,0,0,7,3V2.5A.5.5,0,0,0,6.5,2h-4a.5.5,0,0,0-.5.5v11a.5.5,0,0,0,.5.5h11a.5.5,0,0,0,.5-.5v-4A.5.5,0,0,0,13.5,9Z"
            transform="translate(-2 -1.99)"
          />
          <path
            d="M8.83,7.88a.51.51,0,0,0,.71,0l2.31-2.32,1.28,1.28A.51.51,0,0,0,14,6.49v-4a.52.52,0,0,0-.5-.5h-4A.51.51,0,0,0,9,2.52a.58.58,0,0,0,.14.33l1.28,1.28L8.12,6.46a.51.51,0,0,0,0,.71Z"
            transform="translate(-2 -1.99)"
          />
        </svg>
      </span>
      <span
        class="ExternalLink__VisualHidden-jbkyyd-2 mbLWN"
      >
         (opens in a new tab)
      </span>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "history": Object {
    "action": "POP",
    "block": [Function],
    "canGo": [Function],
    "createHref": [Function],
    "entries": Array [
      Object {
        "hash": "",
        "key": "iok28e",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    ],
    "go": [Function],
    "goBack": [Function],
    "goForward": [Function],
    "index": 0,
    "length": 1,
    "listen": [Function],
    "location": Object {
      "hash": "",
      "key": "iok28e",
      "pathname": "/",
      "search": "",
      "state": undefined,
    },
    "push": [Function],
    "replace": [Function],
  },
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
